File: lib/every/launchd.rb
module Every
  # Everything that touches launchd: plist rendering + launchctl calls.
  # Agents live at ~/Library/LaunchAgents/com.every.<name>.plist and invoke
  # `every run <name>`, which does the actual execution + logging.
  module Launchd
    module_function

    # Common backend interface (see Backend)
    def write(name, schedule)
      write_plist(name, schedule)
    end

    def enable(name)
      bootstrap(name)
    end

    def disable(name)
      bootout(name)
    end

    def unit_path(name)
      plist_path(name)
    end

    # Generic backend capability used by `doctor`.  File-backed schedulers can
    # answer this locally; registry/service-backed schedulers provide their own
    # implementation.
    def resource_exists?(name)
      File.exist?(unit_path(name))
    end

    def delete_units(name)
      p = plist_path(name)
      File.delete(p) if File.exist?(p)
    end

    def label(name)
      "com.every.#{name}"
    end

    def plist_path(name)
      File.join(AGENTS_DIR, "#{label(name)}.plist")
    end

    def uid
      Process.uid
    end

    def write_plist(name, schedule)
      FileUtils.mkdir_p(AGENTS_DIR)
      File.write(plist_path(name), plist_xml(name, schedule))
      plist_path(name)
    end

    def plist_xml(name, schedule)
      agent_log = File.join(LOG_DIR, "_agent.log")
      trigger =
        if schedule.kind == :interval
          "  <key>StartInterval</key>\n" \
          "  <integer>#{schedule.interval}</integer>"
        else
          dicts = schedule.entries.map do |e|
            lines = []
            if e["weekday"]
              lines << "      <key>Weekday</key><integer>#{e['weekday']}</integer>"
            end
            lines << "      <key>Hour</key><integer>#{e['hour']}</integer>"
            lines << "      <key>Minute</key><integer>#{e['minute']}</integer>"
            "    <dict>\n#{lines.join("\n")}\n    </dict>"
          end
          "  <key>StartCalendarInterval</key>\n  <array>\n#{dicts.join("\n")}\n  </array>"
        end

      <<~XML
        <?xml version="1.0" encoding="UTF-8"?>
        <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
        <plist version="1.0">
        <dict>
          <key>Label</key>
          <string>#{xesc(label(name))}</string>
          <key>ProgramArguments</key>
          <array>
            <string>#{xesc(RbConfig.ruby)}</string>
            <string>#{xesc(Runtime.bin)}</string>
            <string>run</string>
            <string>#{xesc(name)}</string>
          </array>
        #{trigger}
        #{env_block}  <key>RunAtLoad</key>
          <false/>
          <key>StandardOutPath</key>
          <string>#{xesc(agent_log)}</string>
          <key>StandardErrorPath</key>
          <string>#{xesc(agent_log)}</string>
        </dict>
        </plist>
      XML
    end

    # Pin the resolved data dir so a launchd-spawned run reads the SAME store the
    # task was added under. launchd doesn't inherit the shell's EVERY_HOME or
    # XDG_DATA_HOME, so without this an XDG (or EVERY_HOME) install's scheduled
    # runs would recompute the default dir, not find the task, and never fire.
    def env_block
      "  <key>EnvironmentVariables</key>\n" \
      "  <dict>\n" \
      "    <key>EVERY_HOME</key><string>#{xesc(DATA_DIR)}</string>\n" \
      "  </dict>\n"
    end

    def xesc(s)
      s.to_s.gsub("&", "&amp;").gsub("<", "&lt;").gsub(">", "&gt;")
    end

    def bootstrap(name)
      path = plist_path(name)
      out, st = Open3.capture2e("launchctl", "bootstrap", "gui/#{uid}", path)
      return true if st.success?

      # Most common failure: an agent with this label is already loaded.
      bootout(name)
      out, st = Open3.capture2e("launchctl", "bootstrap", "gui/#{uid}", path)
      return true if st.success?

      _legacy, st2 = Open3.capture2e("launchctl", "load", "-w", path)
      return true if st2.success?

      raise "launchctl bootstrap failed: #{out.strip}"
    end

    def bootout(name)
      Open3.capture2e("launchctl", "bootout", "gui/#{uid}/#{label(name)}")
    end

    def loaded?(name)
      _out, st = Open3.capture2e("launchctl", "print", "gui/#{uid}/#{label(name)}")
      st.success?
    end

    # All loaded every-agent names in ONE call (vs one `print` per task) — let
    # `list`/`doctor` check membership instead of forking a subprocess per task.
    def loaded_names
      out, st = Open3.capture2e("launchctl", "list")
      return [] unless st.success?
      parse_labels(out)
    end

    # `launchctl list` prints "PID<TAB>Status<TAB>Label" rows; keep our labels.
    def parse_labels(out)
      out.lines.each_with_object([]) do |line, acc|
        lbl = line.split("\t").last.to_s.strip
        acc << lbl.sub("com.every.", "") if lbl.start_with?("com.every.")
      end
    end
  end
end

File: internal/backend/launchd.go
package backend

import (
	"fmt"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"strings"

	"github.com/serhiileniv/every/internal/naming"
	"github.com/serhiileniv/every/internal/schedule"
)

// Launchd schedules through macOS launchd user agents.
type Launchd struct{ cfg Config }

func NewLaunchd(cfg Config) *Launchd { return &Launchd{cfg: cfg} }

func (l *Launchd) Name() string { return "launchd" }

func (l *Launchd) Label(name string) string { return "com.every." + name }

func (l *Launchd) UnitPath(name string) string {
	return filepath.Join(l.cfg.Dirs.Agents, l.Label(name)+".plist")
}

func (l *Launchd) ResourceExists(name string) bool {
	_, err := os.Stat(l.UnitPath(name))
	return err == nil
}

func (l *Launchd) Write(name string, s *schedule.Schedule) error {
	// The last gate before a name becomes a path. See internal/naming: the
	// store is a plain file, so `add`'s sanitizer is not the only way in.
	if err := naming.Validate(name); err != nil {
		return err
	}
	if err := os.MkdirAll(l.cfg.Dirs.Agents, 0o755); err != nil {
		return err
	}
	// logs/ is created here, at add time, so the first fire's _agent.log
	// redirect has somewhere to open. launchd will not create it.
	if err := os.MkdirAll(l.cfg.Dirs.Logs, 0o755); err != nil {
		return err
	}
	return os.WriteFile(l.UnitPath(name), []byte(l.PlistXML(name, s)), 0o644)
}

// PlistXML builds the agent definition.
//
// Built as a string, not through encoding/xml: that package reorders
// attributes, reformats whitespace, and escapes more characters than the three
// entities used here. The output is compared byte for byte against fixtures
// generated from the Ruby, and a plist that differs only cosmetically would
// still make an upgrade rewrite every user's agents for no reason.
func (l *Launchd) PlistXML(name string, s *schedule.Schedule) string {
	// path.Join, not filepath.Join: this string goes INTO a plist, and launchd
	// is macOS-only, so the separator is always "/" regardless of the host
	// building it. filepath.Join uses the HOST's separator, which turned the
	// log path into backslashes when the generator ran on Windows -- harmless
	// in production, since the backend only runs on macOS, but it broke the
	// cross-platform golden tests that exist precisely so these generators are
	// checked everywhere.
	agentLog := path.Join(l.cfg.Dirs.Logs, "_agent.log")

	var trigger string
	if s.Kind == schedule.Interval {
		trigger = "  <key>StartInterval</key>\n  <integer>" + s.Interval.String() + "</integer>"
	} else {
		dicts := make([]string, 0, len(s.Entries))
		for _, e := range s.Entries {
			var lines []string
			// Weekday only when the entry has one: its absence is what makes
			// `day 9am` fire daily rather than on Sunday.
			if e.Weekday != nil {
				lines = append(lines, fmt.Sprintf("      <key>Weekday</key><integer>%d</integer>", *e.Weekday))
			}
			lines = append(lines,
				fmt.Sprintf("      <key>Hour</key><integer>%d</integer>", e.Hour),
				fmt.Sprintf("      <key>Minute</key><integer>%d</integer>", e.Minute))
			dicts = append(dicts, "    <dict>\n"+strings.Join(lines, "\n")+"\n    </dict>")
		}
		trigger = "  <key>StartCalendarInterval</key>\n  <array>\n" +
			strings.Join(dicts, "\n") + "\n  </array>"
	}

	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>%s</string>
  <key>ProgramArguments</key>
  <array>
    <string>%s</string>
    <string>run</string>
    <string>%s</string>
  </array>
%s
%s  <key>RunAtLoad</key>
  <false/>
  <key>StandardOutPath</key>
  <string>%s</string>
  <key>StandardErrorPath</key>
  <string>%s</string>
</dict>
</plist>
`,
		xesc(l.Label(name)),
		xesc(l.cfg.Launcher),
		xesc(name),
		trigger,
		l.envBlock(),
		xesc(agentLog),
		xesc(agentLog))
}

// envBlock pins the resolved data dir.
//
// launchd does not inherit the shell's EVERY_HOME or XDG_DATA_HOME, so without
// this an XDG (or EVERY_HOME) install's scheduled runs would recompute the
// default directory, fail to find the task, and never fire.
func (l *Launchd) envBlock() string {
	return "  <key>EnvironmentVariables</key>\n  <dict>\n" +
		fmt.Sprintf("    <key>EVERY_HOME</key><string>%s</string>\n", xesc(l.cfg.Dirs.Data)) +
		"  </dict>\n"
}

// xesc escapes exactly the three entities the Ruby escaped, in that order.
// Not xml.EscapeText, which also escapes quotes, tabs and newlines as numeric
// references and would change every fixture.
func xesc(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	return strings.ReplaceAll(s, ">", "&gt;")
}

func (l *Launchd) uid() int { return os.Getuid() }

func (l *Launchd) domain() string { return fmt.Sprintf("gui/%d", l.uid()) }

// Enable loads the agent, with a fallback ladder.
//
// bootstrap is the modern API. Its most common failure is an agent with this
// label already loaded, so on failure we bootout and retry. `load -w` is the
// pre-10.10 path and covers hosts where bootstrap is unavailable or refuses
// over a bare SSH session.
func (l *Launchd) Enable(name string) error {
	path := l.UnitPath(name)

	if out, err := runCmd("launchctl", "bootstrap", l.domain(), path); err == nil {
		_ = out
		return nil
	}

	_ = l.Disable(name)

	out, err := runCmd("launchctl", "bootstrap", l.domain(), path)
	if err == nil {
		return nil
	}

	if _, legacyErr := runCmd("launchctl", "load", "-w", path); legacyErr == nil {
		return nil
	}

	// The message carries the second bootstrap's output, which is the
	// informative one; `load -w`'s failure is almost always just "unsupported".
	return fmt.Errorf("launchctl bootstrap failed: %s", strings.TrimSpace(out))
}

// Disable unloads the agent. A task the scheduler no longer has must still
// disable cleanly: CLI rm calls this before deleting the units, so raising on
// an already-absent agent would strand the store entry with no way to remove it.
func (l *Launchd) Disable(name string) error {
	_, _ = runCmd("launchctl", "bootout", l.domain()+"/"+l.Label(name))
	return nil
}

func (l *Launchd) DeleteUnits(name string) error {
	err := os.Remove(l.UnitPath(name))
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

func (l *Launchd) Loaded(name string) bool {
	_, err := runCmd("launchctl", "print", l.domain()+"/"+l.Label(name))
	return err == nil
}

// LoadedNames lists every loaded agent in ONE call, rather than one `print` per
// task -- `list` and `doctor` then test membership instead of forking a
// subprocess for each.
func (l *Launchd) LoadedNames() ([]string, error) {
	out, err := runCmd("launchctl", "list")
	if err != nil {
		// A soft failure: an unreachable launchd means "nothing is loaded",
		// which is what doctor should report rather than a crash.
		return nil, nil
	}
	return parseLabels(out), nil
}

// parseLabels pulls every-owned labels out of `launchctl list` output, whose
// rows are "PID<TAB>Status<TAB>Label".
func parseLabels(out string) []string {
	const prefix = "com.every."
	var names []string
	for _, line := range strings.Split(out, "\n") {
		fields := strings.Split(line, "\t")
		label := strings.TrimSpace(fields[len(fields)-1])
		if strings.HasPrefix(label, prefix) {
			// Replace only the first occurrence, as Ruby's sub did.
			names = append(names, strings.Replace(label, prefix, "", 1))
		}
	}
	return names
}

// runCmd runs a command and returns its merged output.
func runCmd(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	out, err := cmd.CombinedOutput()
	return string(out), err
}

// Render returns what Write would produce, without writing it. Used by the
// migration to decide whether a unit on disk is stale.
func (l *Launchd) Render(name string, s *schedule.Schedule) (string, error) {
	return l.PlistXML(name, s), nil
}
